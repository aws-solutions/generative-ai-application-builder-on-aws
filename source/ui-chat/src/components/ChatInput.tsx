/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { MutableRefObject, useContext, useState } from 'react';
import { Button, Container, StatusIndicator, StatusIndicatorProps, Textarea } from '@cloudscape-design/components';
import HomeContext from '../home/home.context';
import { Message } from '../types/chat';
import { MODEL_MAX_INPUT_LENGTH } from '../utils/constants';
import './ChatInput.css';

interface Props {
    onSend: (message: Message) => void;
    stopConversationRef: MutableRefObject<boolean>;
    socketStatusType: StatusIndicatorProps.Type;
    socketStatusMessage: string;
}

export const ChatInput = ({ onSend, stopConversationRef, socketStatusType, socketStatusMessage }: Props) => {
    const {
        state: { messageIsStreaming, loading }
    } = useContext(HomeContext);

    const [content, setContent] = useState<string>('');

    const handleChange = (value: string) => {
        if (value === '\n') {
            return;
        }
        setContent(value);
    };

    const verifyLength = () => {
        const maxLength = MODEL_MAX_INPUT_LENGTH;

        if (maxLength && content.length > maxLength) {
            alert(`Message limit is ${maxLength} characters. You have entered ${content.length} characters.`);
            return false;
        }

        return true;
    };

    const handleSend = () => {
        if (messageIsStreaming || loading || !verifyLength()) {
            return;
        }
        if (!content) {
            alert('Please enter a message');
            return;
        }
        onSend({ role: 'user', content });
        setContent('');
    };

    const handleKeyDown = (key: string, shiftKey: boolean) => {
        if (key === 'Enter' && !shiftKey) {
            handleSend();
        }
    };

    return (
        <Container>
            <StatusIndicator type={socketStatusType}> {socketStatusMessage} </StatusIndicator>{' '}
            <div className="chatinput-textarea">
                <div style={{ width: '92%', display: 'inline-block' }} data-testid="chat-input-textarea-div">
                    <Textarea
                        data-class-name="chat-input-textarea"
                        data-testid="chat-input-textarea"
                        data-columns="3"
                        placeholder={'Type a message...' || ''}
                        value={content}
                        rows={1}
                        onChange={({ detail }) => handleChange(detail.value)}
                        onKeyDown={({ detail }) => handleKeyDown(detail.key, detail.shiftKey)}
                        autoFocus
                    />
                </div>
                <div
                    style={{
                        float: 'right',
                        display: 'inline-block'
                    }}
                >
                    <Button
                        iconName={messageIsStreaming || loading ? 'status-in-progress' : 'check'}
                        loading={messageIsStreaming || loading}
                        variant="icon"
                        onClick={handleSend}
                        disabled={messageIsStreaming || loading}
                        data-testid={'send-button'}
                    />
                </div>
            </div>
        </Container>
    );
};
