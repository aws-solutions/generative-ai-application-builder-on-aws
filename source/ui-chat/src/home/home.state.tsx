/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Conversation } from '../types/chat';
import { MAX_PROMPT_TEMPLATE_LENGTH, MODEL_MAX_INPUT_LENGTH } from '../utils/constants';

export interface UseCaseConfigType {
    KnowledgeBaseParams: {
        ReturnSourceDocs: boolean;
        KnowledgeBaseType: string;
        KendraKnowledgeBaseParams: {
            ExistingKendraIndexId: string;
            RoleBasedAccessControlEnabled: boolean;
        };
        NumberOfDocs: number;
        ScoreThreshold: number;
        NoDocsFoundResponse: string;
    };
    ConversationMemoryParams: {
        HumanPrefix: string;
        ConversationMemoryType: string;
        ChatHistoryLength: number;
        AiPrefix: string;
    };
    UseCaseName: string;
    LlmParams: {
        Streaming: boolean;
        Temperature: number;
        Verbose: boolean;
        BedrockLlmParams: {
            ModelId: string;
        };
        ModelProvider: string;
        PromptParams: {
            UserPromptEditingEnabled: boolean;
            DisambiguationEnabled: boolean;
            MaxInputTextLength: number;
            RephraseQuestion: boolean;
            PromptTemplate: string;
            MaxPromptTemplateLength: number;
            DisambiguationPromptTemplate: string;
        };
        ModelParams: object;
        RAGEnabled: boolean;
    };
}

export interface HomeInitialState {
    loading: boolean;
    messageIsStreaming: boolean;
    selectedConversation: Conversation | undefined;
    promptTemplate: string;
    defaultPromptTemplate: string;
    RAGEnabled: boolean;
    useCaseConfig: UseCaseConfigType;
    userPromptEditingEnabled: boolean;
    maxPromptTemplateLength: number;
    maxInputTextLength: number;
}

export const initialState: HomeInitialState = {
    loading: false,
    messageIsStreaming: false,
    selectedConversation: {
        id: '',
        name: 'New Conversation',
        messages: []
    },
    promptTemplate: '',
    defaultPromptTemplate: '',
    RAGEnabled: false,
    useCaseConfig: {} as UseCaseConfigType,
    userPromptEditingEnabled: true,
    maxPromptTemplateLength: MAX_PROMPT_TEMPLATE_LENGTH,
    maxInputTextLength: MODEL_MAX_INPUT_LENGTH
};
