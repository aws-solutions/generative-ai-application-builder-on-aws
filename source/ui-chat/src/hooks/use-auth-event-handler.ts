// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { useCallback, useEffect } from 'react';
import { fetchAuthSession, Hub } from '@aws-amplify/core';
import { signOut, signInWithRedirect } from '@aws-amplify/auth';
import { useSelector } from 'react-redux';
import { RootState } from '../store/store';

/**
 * Interface defining the return type of the useAuthEventHandler hook
 * @property getAccessToken - Function to retrieve the current access token
 * @property handleSignIn - Function to initiate the sign in flow
 * @property handleSignOut - Function to sign out the current user
 */
interface UseAuthEventHandlerReturn {
    getAccessToken: () => Promise<string>;
    handleSignIn: () => Promise<void>;
    handleSignOut: () => Promise<void>;
}

/**
 * Custom hook to handle authentication events and token management
 * @param checkUser - Function to verify the current user's authentication status
 * @param resetUserState - Function to clear user state on sign out or auth failure
 * @returns {UseAuthEventHandlerReturn}
 */
export const useAuthEventHandler = (
    checkUser: () => Promise<void>,
    resetUserState: () => void
): UseAuthEventHandlerReturn => {
    const runtimeConfig = useSelector((state: RootState) => state.config.runtimeConfig);
    const userPoolClientId = runtimeConfig?.UserPoolClientId;

    /**
     * Retrieves the current access token from the auth session
     * @returns Promise resolving to the access token string
     * @throws Error if no access token is available
     */
    const getAccessToken = useCallback(async (): Promise<string> => {
        try {
            const session = await fetchAuthSession();
            const accessToken = session.tokens?.accessToken?.toString();
            if (!accessToken) {
                throw new Error('No access token available');
            }
            return accessToken;
        } catch (error) {
            console.error('Error getting access token:', error);
            throw error;
        }
    }, []);

    /**
     * Initiates the sign in flow using redirect
     */
    const handleSignIn = async () => {
        try {
            await signInWithRedirect();
        } catch (error) {
            console.error('Error starting sign in:', error);
        }
    };

    /**
     * Signs out the current user and resets the user state
     */
    const handleSignOut = async () => {
        try {
            await signOut();
            resetUserState();
        } catch (error) {
            console.error('Error signing out:', error);
        }
    };

    /**
     * Sets up an auth event listener to handle various authentication state changes
     */
    useEffect(() => {
        const listener = Hub.listen('auth', ({ payload }) => {
            const eventHandlers: Record<string, () => void> = {
                signedIn: checkUser,
                signedOut: resetUserState,
                tokenRefresh: checkUser,
                tokenRefresh_failure: resetUserState
            };

            const handler = eventHandlers[payload.event];
            if (handler) {
                handler();
            }
        });

        return () => listener();
    }, [checkUser, resetUserState]);

    // Cross-tab auth state listener
    useEffect(() => {
        if (!userPoolClientId) return;

        const handleStorageChange = async (event: StorageEvent) => {
            const lastAuthUserKey = `CognitoIdentityServiceProvider.${userPoolClientId}.LastAuthUser`;
            const accessTokenKey = `CognitoIdentityServiceProvider.${userPoolClientId}`;

            const isSignOutEvent =
                (event.key === lastAuthUserKey && event.newValue === null) ||
                (event.key?.startsWith(accessTokenKey) &&
                    event.key?.includes('accessToken') &&
                    event.newValue === null);

            if (isSignOutEvent) {
                try {
                    // Clean up local state
                    resetUserState();
                    await signOut();
                } catch (error) {
                    console.error('Error during cross-tab signout:', error);
                }
            }
        };

        window.addEventListener('storage', handleStorageChange);
        return () => window.removeEventListener('storage', handleStorageChange);
    }, [userPoolClientId, resetUserState]);

    return { getAccessToken, handleSignIn, handleSignOut };
};
