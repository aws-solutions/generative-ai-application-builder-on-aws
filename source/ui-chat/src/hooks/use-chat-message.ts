// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { useCallback, useReducer } from 'react';
import { ChatResponse, isChatSuccessResponse, isErrorResponse, SourceDocument } from '../models';
import { ChatActionTypes, Message } from '../pages/chat/types';
import { useUser } from '../contexts/UserContext';
import { chatReducer } from '../reducers/chat-reducer';
import { END_CONVERSATION_TOKEN } from '../utils/constants';

/**
 * Interface representing the state of a chat conversation
 * @property messages - Array of chat messages
 * @property currentResponse - Current response being generated
 * @property isGenAiResponseLoading - Flag indicating if AI is generating a response
 * @property sourceDocuments - Array of source documents referenced in the conversation
 * @property conversationId - Unique identifier for the conversation
 */
export interface ChatState {
    messages: Message[];
    currentResponse: string;
    isGenAiResponseLoading: boolean;
    sourceDocuments: SourceDocument[];
    conversationId: string;
}

/**
 * Initial state for the chat
 */
const initialState: ChatState = {
    messages: [],
    currentResponse: '',
    isGenAiResponseLoading: false,
    sourceDocuments: [],
    conversationId: ''
};

/**
 * Custom hook to manage chat messages and state
 * @returns {Object} Chat state and handler functions
 */
export const useChatMessages = () => {
    const [state, dispatch] = useReducer(chatReducer, initialState);
    const { userId } = useUser();

    /**
     * Handles incoming chat responses and updates state accordingly
     * @param {ChatResponse} response - The chat response to process
     */
    const handleMessage = useCallback(
        (response: ChatResponse) => {
            try {
                if (!response) {
                    console.error('Received invalid response');
                    return;
                }

                if (isErrorResponse(response) && response.errorMessage) {
                    dispatch({
                        type: ChatActionTypes.SET_ERROR,
                        payload: response.errorMessage
                    });
                }

                if (isChatSuccessResponse(response) && response.conversationId) {
                    dispatch({ type: ChatActionTypes.SET_CONVERSATION_ID, payload: response.conversationId });
                }

                if (isChatSuccessResponse(response) && response.sourceDocument) {
                    dispatch({ type: ChatActionTypes.ADD_SOURCE_DOCUMENT, payload: response.sourceDocument });
                }

                if (isChatSuccessResponse(response) && response.rephrased_query) {
                    dispatch({ type: ChatActionTypes.ADD_REPHRASED_QUERY, payload: response.rephrased_query });
                }

                if (response.data !== undefined) {
                    if (response.data === END_CONVERSATION_TOKEN) {
                        dispatch({ type: ChatActionTypes.COMPLETE_AI_RESPONSE });
                    } else {
                        // Send only the new data token, not the concatenated response
                        dispatch({ 
                            type: ChatActionTypes.UPDATE_AI_RESPONSE, 
                            payload: { 
                                content: response.data,
                                messageId: response.messageId 
                            } 
                        });
                    }
                }
            } catch (error) {
                dispatch({
                    type: ChatActionTypes.SET_ERROR,
                    payload: 'An error occurred while processing the message.'
                });
            }
        },
        [state.currentResponse]
    );

    /**
     * Resets the chat state to initial values
     */
    const resetChat = useCallback(() => {
        dispatch({ type: ChatActionTypes.RESET_CHAT });
    }, []);

    /**
     * Updates the messages array in the chat state
     * @param {Message[]} messages - Array of messages to set
     */
    const handleSetMessages = (messages: Message[]) => {
        dispatch({ type: ChatActionTypes.SET_MESSAGES, payload: messages });
    };

    /**
     * Adds a new user message to the chat
     * @param {string} userInput - The message content from the user
     */
    const handleAddUserMessage = (userInput: string) => {
        dispatch({ type: ChatActionTypes.ADD_USER_MESSAGE, payload: { content: userInput, authorId: userId } });
    };

    return {
        messages: state.messages,
        currentResponse: state.currentResponse,
        isGenAiResponseLoading: state.isGenAiResponseLoading,
        sourceDocuments: state.sourceDocuments,
        conversationId: state.conversationId,
        handleMessage,
        addUserMessage: handleAddUserMessage,
        resetChat,
        setMessages: handleSetMessages,
        setIsGenAiResponseLoading: (loading: boolean) => {
            if (!loading) {
                dispatch({ type: ChatActionTypes.COMPLETE_AI_RESPONSE });
            }
        }
    };
};
