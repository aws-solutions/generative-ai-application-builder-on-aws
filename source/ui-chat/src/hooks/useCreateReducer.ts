/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { useMemo, useReducer } from 'react';

// Extracts property names from initial state of reducer to allow typesafe dispatch objects
export type FieldNames<T> = {
    [K in keyof T]: T[K] extends string ? K : K;
}[keyof T];

// Returns the Action Type for the dispatch object to be used for typing in things like context
export type ActionType<T> = { type: 'reset' } | { type?: 'change'; field: FieldNames<T>; value: any };

// Returns a typed dispatch and state
export const useCreateReducer = <T>({ initialState }: { initialState: T }) => {
    type Action = { type: 'reset' } | { type?: 'change'; field: FieldNames<T>; value: any };

    const reducer = (state: T, action: Action) => {
        if (!action.type) return { ...state, [action.field]: action.value };

        if (action.type === 'reset') return initialState;

        throw new Error();
    };

    const [state, dispatch] = useReducer(reducer, initialState);

    return useMemo(() => ({ state, dispatch }), [state, dispatch]);
};
