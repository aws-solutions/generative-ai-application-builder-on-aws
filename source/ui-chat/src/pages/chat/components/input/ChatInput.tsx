// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { FormField, Link, PromptInput } from '@cloudscape-design/components';
import { memo, useCallback, useState } from 'react';
import {
    CHAT_INPUT_MAX_ROWS,
    CONSTRAINT_TEXT_ERROR_COLOR,
    DEFAULT_CHAT_INPUT_MAX_LENGTH,
    DOCS_LINKS
} from '../../../../utils/constants';
import { useSelector } from 'react-redux';
import { RootState } from '../../../../store/store';
import { getMaxInputTextLength } from '../../../../store/configSlice';
import { formatCharacterCount } from '../../../../utils/validation';

/**
 * Props interface for the ChatInput component
 * @interface ChatInputProps
 * @property {boolean} isLoading - Flag indicating if a request is in progress
 * @property {function} onSend - Callback function triggered when a message is sent
 */
interface ChatInputProps {
    isLoading: boolean;
    onSend: (value: string) => void;
}

/**
 * ChatInput component that renders a form field with a prompt input
 * @param {ChatInputProps} props - Component props
 * @param {boolean} props.isLoading - Flag indicating if a request is in progress
 * @param {function} props.onSend - Callback function triggered when a message is sent
 * @returns {JSX.Element} Rendered form field with prompt input
 */
export const ChatInput = memo(({ isLoading, onSend }: ChatInputProps) => {
    const [inputText, setInputText] = useState('');

    // Selector to determine if user is internal
    const isInternalUser = useSelector((state: RootState) => state.config.runtimeConfig?.IsInternalUser) === 'true';

    const maxInputLength = useSelector((state: RootState) => {
        try {
            return getMaxInputTextLength(state);
        } catch {
            return DEFAULT_CHAT_INPUT_MAX_LENGTH;
        }
    });

    /**
     * Handles the action when user submits input
     * @param {Object} param - Event parameter object
     * @param {Object} param.detail - Contains the input value
     */
    const handleAction = useCallback(
        ({ detail }: { detail: { value: string } }) => {
            if (!detail.value?.trim() || isLoading) return;
            // Only send if the text is within the limit
            if (detail.value.length <= maxInputLength) {
                onSend(detail.value);
                setInputText('');
            }
        },
        [isLoading, onSend, maxInputLength]
    );

    const characterCount = inputText.length;
    const isOverLimit = characterCount > maxInputLength;

    return (
        <FormField
            stretch
            constraintText={
                <>
                    <span style={{ color: isOverLimit ? CONSTRAINT_TEXT_ERROR_COLOR : 'inherit' }}>
                        {characterCount}/{formatCharacterCount(maxInputLength)} characters.{' '}
                    </span>
                    {isInternalUser && (
                        <>
                            Use of this service is subject to the{' '}
                            <Link href={DOCS_LINKS.GEN_AI_POLICY} external variant="primary" fontSize="inherit">
                                Third Party Generative AI Use Policy
                            </Link>
                            .
                        </>
                    )}
                </>
            }
        >
            <PromptInput
                onChange={({ detail }) => {
                    // Allow any length of input
                    setInputText(detail.value);
                }}
                onAction={handleAction}
                value={inputText}
                actionButtonAriaLabel={
                    isLoading
                        ? 'Send message button - suppressed'
                        : isOverLimit
                          ? 'Cannot send - message too long'
                          : 'Send message'
                }
                actionButtonIconName="send"
                ariaLabel={isLoading ? 'Chat input text - suppressed' : 'Chat input text'}
                placeholder="Ask a question"
                autoFocus
                maxRows={CHAT_INPUT_MAX_ROWS}
                data-testid="chat-input"
            />
        </FormField>
    );
});
