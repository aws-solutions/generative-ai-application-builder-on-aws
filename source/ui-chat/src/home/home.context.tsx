/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Dispatch, createContext } from 'react';

import { ActionType } from '../hooks/useCreateReducer';

import { Conversation } from '../types/chat';
import { KeyValuePair } from '../types/misc';

import { HomeInitialState } from './home.state';

export interface HomeContextProps {
    state: HomeInitialState;
    dispatch: Dispatch<ActionType<HomeInitialState>>;
    handleUpdateConversation: (conversation: Conversation, data: KeyValuePair) => void;
}

const HomeContext = createContext<HomeContextProps>(undefined!);

export default HomeContext;
