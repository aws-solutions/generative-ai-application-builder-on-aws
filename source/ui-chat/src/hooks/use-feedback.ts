// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { useState } from 'react';
import { useSelector } from 'react-redux';
import { useSubmitFeedbackMutation } from '../store/solutionApi';
import { getUseCaseConfigKey, getUseCaseId } from '../store/configSlice';
import { FeedbackFormData, FeedbackType } from '../pages/chat/components/input/FeedbackForm';
import { ChatBubbleMessage } from '../pages/chat/types';

export interface UseFeedbackResult {
    showFeedbackForm: boolean;
    setShowFeedbackForm: (show: boolean) => void;
    feedbackType: FeedbackType;
    setFeedbackType: (type: FeedbackType) => void;
    feedbackSubmitted: boolean;
    feedbackError: string | null;
    isSubmittingFeedback: boolean;
    handleFeedbackButtonClick: (type: FeedbackType) => void;
    handleFeedbackSubmit: (feedbackData: FeedbackFormData) => void;
}

/**
 * Custom hook to manage feedback state and submission
 * @param message The message to provide feedback for
 * @param onFeedback Callback function when feedback is submitted
 * @returns Object containing feedback state and handlers
 */
export const useFeedback = (message: ChatBubbleMessage, conversationId: string): UseFeedbackResult => {
    const [showFeedbackForm, setShowFeedbackForm] = useState(false);
    const [feedbackType, setFeedbackType] = useState<FeedbackType>('');
    const [feedbackSubmitted, setFeedbackSubmitted] = useState(false);
    const [feedbackError, setFeedbackError] = useState<string | null>(null);
    const [submitFeedback, { isLoading: isSubmittingFeedback }] = useSubmitFeedbackMutation();
    const useCaseConfigKey = useSelector(getUseCaseConfigKey);
    const useCaseId = useSelector(getUseCaseId);

    const handleFeedbackButtonClick = (type: FeedbackType) => {
        setFeedbackType(type);
        setShowFeedbackForm(true);
    };

    const handleFeedbackSubmit = (feedbackData: FeedbackFormData) => {
        if (!useCaseId) {
            console.error('Use case ID not found in runtime config');
            return;
        }
        if (!useCaseConfigKey) {
            console.error('Use case config record key not found in runtime config');
            return;
        }

        if (!message.messageId) {
            console.error('Message ID not found for feedback submission');
            setFeedbackError('Unable to submit feedback: Message ID not found');
            return;
        }
        
        const feedbackPayload = {
            useCaseRecordKey: useCaseConfigKey,
            conversationId: conversationId,
            messageId: message.messageId,
            feedback: feedbackType === 'helpful' ? 'positive' : 'negative',
            feedbackReason: feedbackData.reasons,
            comment: feedbackData.comment,
            ...(message.rephrasedQuery && { rephrasedQuery: message.rephrasedQuery }),
            ...((message.sourceDocuments && message.sourceDocuments.length > 0) && { 
                sourceDocuments: message.sourceDocuments
                    .map(doc => doc.location)
                    .filter(Boolean) // Remove any undefined or null values
                    .filter((value, index, self) => self.indexOf(value) === index) // Remove duplicates
            }),
        };

        submitFeedback({
            useCaseId,
            feedbackData: feedbackPayload
        })
            .unwrap()
            .then(() => {
                setShowFeedbackForm(false);
                setFeedbackSubmitted(true);
                setFeedbackError(null);
            })
            .catch((error) => {
                console.error('Error submitting feedback:', error);
                setShowFeedbackForm(false);
                setFeedbackSubmitted(false);
                setFeedbackError(error.data?.message || 'Failed to submit feedback. Please try again or contact a system administrator.');
            });
    };

    return {
        showFeedbackForm,
        setShowFeedbackForm,
        feedbackType,
        setFeedbackType,
        feedbackSubmitted,
        feedbackError,
        isSubmittingFeedback,
        handleFeedbackButtonClick,
        handleFeedbackSubmit
    };
};
