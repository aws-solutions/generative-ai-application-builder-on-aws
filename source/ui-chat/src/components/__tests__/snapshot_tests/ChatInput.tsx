/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 **********************************************************************************************************************/

import { Dispatch } from 'react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import HomeContext from '../../../home/home.context';
import { HomeInitialState } from '../../../home/home.state';
import { ActionType } from '../../../hooks/useCreateReducer';
import { ChatInput } from '../../ChatInput';

jest.mock('@cloudscape-design/components', () => {
    const Components = jest.genMockFromModule('@cloudscape-design/components') as any;
    for (const componentName of Object.keys(Components)) {
        Components[componentName] = componentName;
    }
    return Components;
});

const createMockRefObject = (initialValue: boolean) => ({
    current: initialValue
});

describe('Chat', () => {
    const contextValue = {
        dispatch: jest.fn() as Dispatch<ActionType<HomeInitialState>>,
        state: {
            loading: false,
            messageIsStreaming: false,
            selectedConversation: {
                id: '',
                name: 'New Conversation',
                messages: []
            },
            promptTemplate: '',
            defaultPromptTemplate: ''
        },
        handleUpdateConversation: jest.fn()
    };

    test('Snapshot test for Chat', async () => {
        const stopConersationRef = createMockRefObject(false);
        const tree = renderer
            .create(
                <>
                    <HomeContext.Provider
                        value={{
                            ...contextValue
                        }}
                    >
                        <ChatInput
                            onSend={jest.fn()}
                            stopConversationRef={stopConersationRef}
                            socketStatusType={'success'}
                            socketStatusMessage="Connected"
                        />
                    </HomeContext.Provider>
                </>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
