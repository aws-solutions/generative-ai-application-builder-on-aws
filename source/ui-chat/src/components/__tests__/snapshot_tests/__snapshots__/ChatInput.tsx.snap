// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Snapshot test for Chat 1`] = `
<Container>
  <StatusIndicator
    type="success"
  >
     
    Connected
     
  </StatusIndicator>
   
  <div
    className="chatinput-textarea"
  >
    <div
      data-testid="chat-input-textarea-div"
      style={
        Object {
          "display": "inline-block",
          "width": "92%",
        }
      }
    >
      <Textarea
        autoFocus={true}
        data-class-name="chat-input-textarea"
        data-columns="3"
        data-testid="chat-input-textarea"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Type a message..."
        rows={1}
        value=""
      />
    </div>
    <div
      style={
        Object {
          "display": "inline-block",
          "float": "right",
        }
      }
    >
      <Button
        data-testid="send-button"
        disabled={false}
        iconName="check"
        loading={false}
        onClick={[Function]}
        variant="icon"
      />
    </div>
  </div>
</Container>
`;
